# Parcial 1 Minería de datos
## Carlos Fernando Gonzalzez Sarmiento

In [45]:

!pip install dfply

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 612 kB 7.4 MB/s 


In [46]:
import pandas as pd
import numpy as np
import csv
from dfply import *


In [2]:
pd.set_option('display.max_columns', None) #Mostrar todas las columnas


In [3]:
df = pd.read_csv(
    'https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/student-mat.csv',
    sep = ";")
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  
0     1       3         6   5   6   6  
1     1       3         4   5   5   6  
2     3       3        10   7   8  10  
3     1       5         2  15  14  15  
4     2       5         4   6  10  10

# Punto 1

In [4]:
df.isna().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

In [5]:
df.isnull().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

## R\\\

Basandonos en los anteriores codigos, no existe ningun valor vacío o nulo para alguna de las variables de la base.

# Punto 2

In [9]:
(df['address']=='U').sum()

307

In [11]:
(df['address']=='R').sum()

88

In [13]:
(df['address']=='R').sum()+(df['address']=='U').sum()

395

In [77]:
df.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  395.000000  395.000000  395.000000  395.000000  
mean     5.708861   10.908861   10.713924   10.415190  
std      8.003096    3.319195    3.761505    4.581443  
min      0.000000    3.000000    0.000000    0.000000  
25%      0.000000    8.000000    9.000000    8.000000  
50%      4.000000   11.000000   11.000000   11.000000  
75%      8.000000   13.000000   13.000000   14.000000  
max     75.000000   19.000000   19.000000   20.000000

In [14]:
a=df['address']=='R'

In [18]:
df[a].shape

(88, 33)

In [21]:
df[a]

school sex  age address famsize Pstatus  Medu  Fedu      Mjob     Fjob  \
24      GP   F   15       R     GT3       T     2     4  services   health   
32      GP   M   15       R     GT3       T     4     3   teacher  at_home   
37      GP   M   16       R     GT3       A     4     4     other  teacher   
38      GP   F   15       R     GT3       T     3     4  services   health   
39      GP   F   15       R     GT3       T     2     2   at_home    other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...      ...   
385     MS   F   18       R     GT3       T     2     2   at_home    other   
386     MS   F   18       R     GT3       T     4     4   teacher  at_home   
387     MS   F   19       R     GT3       T     2     3  services    other   
392     MS   M   21       R     GT3       T     1     1     other    other   
393     MS   M   18       R     LE3       T     3     2  services    other   

         reason guardian  traveltime  studytime  failures schoolsup famsup  \
24       course   mother           1          3         0       yes    yes   
32       course   mother           1          2         0        no    yes   
37   reputation   mother           2          3         0        no    yes   
38       course   mother           1          3         0       yes    yes   
39   reputation   mother           1          1         0       yes    yes   
..          ...      ...         ...        ...       ...       ...    ...   
385       other   mother           2          3         0        no     no   
386  reputation   mother           3          1         0        no    yes   
387      course   mother           1          3         1        no     no   
392      course    other           1          1         3        no     no   
393      course   mother           3          1         0        no     no   

    paid activities nursery higher internet romantic  famrel  freetime  goout  \
24   yes        yes     yes    yes      yes       no       4         3      2   
32    no        yes     yes    yes      yes      yes       4         5      2   
37    no        yes     yes    yes      yes      yes       2         4      3   
38   yes        yes     yes    yes      yes       no       4         3      2   
39   yes        yes     yes    yes       no       no       4         3      1   
..   ...        ...     ...    ...      ...      ...     ...       ...    ...   
385  yes         no     yes    yes       no       no       5         3      3   
386  yes        yes     yes    yes      yes      yes       4         4      3   
387   no        yes      no    yes      yes       no       5         4      2   
392   no         no      no    yes       no       no       5         5      3   
393   no         no      no    yes      yes       no       4         4      1   

     Dalc  Walc  health  absences  G1  G2  G3  
24      1     1       5         2  10   9   8  
32      1     1       5         0  17  16  16  
37      1     1       5         7  15  16  15  
38      1     1       5         2  12  12  11  
39      1     1       2         8  14  13  13  
..    ...   ...     ...       ...  ..  ..  ..  
385     1     3       4         2  10   9  10  
386     2     2       5         7   6   5   6  
387     1     2       5         0   7   5   0  
392     3     3       3         3  10   8   7  
393     3     4       5         0  11  12  10  

[88 rows x 33 columns]

In [86]:
Base88 = (df>>
  mask(X.sex=='M',X.address=='R'))
Base88.shape# Colegio Mousinho con internet

(44, 33)

In [25]:
df2=df[a]
(df2['sex']=='M').sum()

44

## R\\\

Para la zona urbana viven 307 estudiantes, para la zona rural viven 88 estudiantes.

44 estudiantes rurales son hombres.

# Punto 3

In [37]:
(df2['absences']>5).sum()

38

In [42]:
a2=df2['absences']>5
a2

24     False
32     False
37      True
38     False
39      True
       ...  
385    False
386     True
387    False
392    False
393    False
Name: absences, Length: 88, dtype: bool

In [59]:
Base2 = (df >>
 mask(X.sex=='M',X.address=='R',X.absences>5))

In [60]:
Base2.shape

(18, 33)

## R\\\

38 estudiantes de la zona rural tienen mas de 5 fallas.

Y 18 de ellos son hombres

# Punto 4

In [65]:
Base3 = (df >>
  select(X.school,X.G3,X.internet))
Base3

school  G3 internet
0       GP   6       no
1       GP   6      yes
2       GP  10      yes
3       GP  15      yes
4       GP  10       no
..     ...  ..      ...
390     MS   9       no
391     MS  16      yes
392     MS   7       no
393     MS  10      yes
394     MS   9      yes

[395 rows x 3 columns]

In [72]:
Base4 = (Base3 >>
  mask(X.school=='GP',X.internet=='yes'))
Base4['G3'].mean()# Colegio Gabriel Pereira con internet

10.632996632996633

In [73]:
Base5 = (Base3 >>
  mask(X.school=='GP',X.internet=='no'))
Base5['G3'].mean()# Colegio Gabriel Pereira sin internet

9.673076923076923

In [75]:
Base6 = (Base3 >>
  mask(X.school=='MS',X.internet=='yes'))
Base6['G3'].mean()# Colegio Mousinho con internet

10.46875

In [76]:
Base7 = (Base3 >>
  mask(X.school=='MS',X.internet=='no'))
Base7['G3'].mean()# Colegio Mousinho sin internet

8.428571428571429

## R\\\

La media de las notas de matematicas de los estudiantes del Colegio Gabriel Pereira que poseen internet, es: 10.632996632996633.

La media de las notas de matematicas de los estudiantes del Colegio Gabriel Pereira que no poseen internet, es: 9.673076923076923.

La media de las notas de matematicas de los estudiantes del Colegio Mousinho que poseen internet, es: 10.46875.

La media de las notas de matematicas de los estudiantes del Colegio Mousinho que no poseen internet, es: 8.428571428571429.

# Punto 5

In [78]:
Base8 = (df >>
 mutate(
   calf_ttl = X.G1*0.3+X.G2*0.3+X.G3*0.4

 ))
Base8.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  calf_ttl  
0     1       3         6   5   6   6       5.7  
1     1       3         4   5   5   6       5.4  
2     3       3        10   7   8  10       8.5  
3     1       5         2  15  14  15      14.7  
4     2       5         4   6  10  10       8.8

In [80]:
Base8.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3    calf_ttl  
count  395.000000  395.000000  395.000000  395.000000  395.000000  
mean     5.708861   10.908861   10.713924   10.415190   10.652911  
std      8.003096    3.319195    3.761505    4.581443    3.769058  
min      0.000000    3.000000    0.000000    0.000000    1.200000  
25%      0.000000    8.000000    9.000000    8.000000    8.300000  
50%      4.000000   11.000000   11.000000   11.000000   10.700000  
75%      8.000000   13.000000   13.000000   14.000000   13.300000  
max     75.000000   19.000000   19.000000   20.000000   19.400000

In [84]:
Base8['calf_ttl'].median()

10.7

## R\\\

La media calificacion total es: 10.652911

La mediana calificacion total es: 10.7

La mínima calificación total es: 1.2

La máxima calificación total es: 19.4

# Punto 6

En mi opinión realizaría un análisis de regresión, para predecir, usaría la variable de G3, que corresponde a la nota de tercer corte, ya que es la última y a partir de variables como las notas de segundo y primer corte, además de a que colegio pertenece y entre otras, se podría predecir la ultima nota.